# Metisa - Data Science Challenge (Calvin)

Data: Anonymous Microsoft Web Data Data Set - http://archive.ics.uci.edu/ml/datasets/Anonymous+Microsoft+Web+Data

The data is in an ASCII-based sparse-data format called "DST". Each line of the data file starts with a letter which tells the line's type. The three line types of interest are Attribute, Case and Vote. Each Attribute is a website, each Case is a user and each Vote is an Attribute that the user visited. For more details, please read the data description file for the structure of the data set.

##Task:
1. Assuming we are at a time such that we only have the training data, we want to recommend websites that the users should visit based on their user ID (case ID number). Please construct a recommender system, train it with the training data set and then conduct recommendation for the users given the user ID.
2. Please also write the procedure to test your recommender with the test data set. Explain the metrics that you use for the testing.

##Please also answer the following questions:
1. What are the pros and cons of the recommendation algorithm you have used?
2. How did you evaluate your recommender's performance? Why?
3. Are you happy with you recommender's results? What could be a suitable baseline to compare your classifier's performance to?

##Remarks:
1. The challenge does not have 'the one' solution or answer. There are many ways to approach the task. Same holds true for the accompanying questions. Please motivate all the choices you have made.
2. We have stated the task with many implicit and explicit requirements. If you cannot comply with any of these requirements, please state this and work around.
3. We also value your input on how this challenge can be improved.
4. Very important: we want to see how you think. Please write down all your thoughts, however preliminary. We much prefer that you discuss an issue without offering a solution, rather than not mentioning it.


## Relevant Information:

    We created the data by sampling and processing the www.microsoft.com logs.
    The data records the use of www.microsoft.com by 38000 anonymous,
    randomly-selected users. For each user, the data lists all the areas of
    the web site (Vroots) that user visited in a one week timeframe.

    Users are identified only by a sequential number, for example, User #14988,
    User #14989, etc. The file contains no personally identifiable information.
    The 294 Vroots are identified by their title (e.g. "NetShow for PowerPoint")
    and URL (e.g. "/stream"). The data comes from one week in February, 1998.

    Dataset format:
	-- The data is in an ASCII-based sparse-data format called "DST".
           Each line of the data file starts with a letter which tells the line's type.
           The three line types of interest are:
               -- Attribute lines:
		     For example, 'A,1277,1,"NetShow for PowerPoint","/stream"'
                     Where:
                        'A' marks this as an attribute line,
                        '1277' is the attribute ID number for an area of the website
                                 (called a Vroot),
	                '1' may be ignored,
			'"NetShow for PowerPoint"' is the title of the Vroot,
                        '"/stream"' is the URL relative to "http://www.microsoft.com"
                -- Case and Vote Lines:
                    For each user, there is a case line followed by zero or more vote lines.
                     For example:
                           C,"10164",10164
                           V,1123,1
                           V,1009,1
                           V,1052,1
                      Where:
                         'C' marks this as a case line,
                          '10164' is the case ID number of a user,
                         'V' marks the vote lines for this case,
                         '1123', 1009', 1052' are the attributes ID's of Vroots that a
                                user visited.
                          '1' may be ignored.

## Number of Instances:
      -- Training: 32711
      -- Testing:   5000
    Each instance represents an anonymous, randomly selected user of the web site.

## Number of Attributes: 294

## Attribute Information:
   Each attribute is an area ("vroot") of the www.microsoft.com web site.

   The datasets record which Vroots each user visited in a one-week timeframe
   in Feburary 1998.

## Missing Attribute Values: The data is very sparse, so vroot visits are explicit,
    nonvisits are implicit (missing).

## Class Distribution: 
    Mean number of vroot visits per case: 3.0

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_table('anonymous-msweb.data', skiprows = 7, sep = ',', header = None, names = ['Attribute', 'ID', 'Ig', 'Vroot', 'URL'])

/home/calvin/anaconda3/lib/python3.4/site-packages/pandas/io/parsers.py:1170: DtypeWarning: Columns (3,4) have mixed types. Specify dtype option on import or set low_memory=False.
  data = self._reader.read(nrows)


## First, divide the data into two tables: 
### 1) Attribute table - store all the attribute information
### 2) Cases and Votes tables - for extract information from our cases

In [21]:
# Attribute table
dfA = data.ix[0:293,:]
dfA.head()

Attribute    ID  Ig                            Vroot          URL
0         A  1287   1          International AutoRoute   /autoroute
1         A  1288   1                          library     /library
2         A  1289   1  Master Chef Product Information  /masterchef
3         A  1297   1                  Central America    /centroam
4         A  1215   1         For Developers Only Info   /developer

In [12]:
# The Case and Vote data is then store in dfC
dfC = data.ix[294:,:3]
dfC = dfC.reset_index()
dfC.head()

index Attribute     ID     Ig
0    294         C  10001  10001
1    295         V   1000      1
2    296         V   1001      1
3    297         V   1002      1
4    298         C  10002  10002

In [20]:
dfA[dfA['ID'].isin([1000,1001,1002])]

Attribute    ID  Ig                   Vroot       URL
78          A  1001   1         Support Desktop  /support
217         A  1002   1  End User Produced View   /athome
268         A  1000   1                  regwiz   /regwiz

# Strategy:
1. Find relationships between Attribute, i.e. in the first case, ID:10001 visit site 1000, then visit 1001 and 1002, so it indicate that there is a relation between Support Desktop, End User Produced View and regwiz.
2. Formulate these relations from the case and vote data, then we can use these information to recommend website based on the history of website visited.


In [33]:
test = np.empty(20)*np.nan

for i in range(0,10):
    if dfC.Attribute[i] == 'C':
        test[i] = dfC.ID[i+1]
test

array([ 1000.,    nan,    nan,    nan,  1001.,    nan,    nan,  1001.,
          nan,    nan,    nan,    nan,    nan,    nan,    nan,    nan,
          nan,    nan,    nan,    nan])

In [34]:
test = np.empty(10,2)*np.nan
test

TypeError: data type not understood

In [5]:
dfC = data.loc[data['Attribute'] == 'C']
dfC = dfC.reset_index()
#dfC['Nvote'] = float('NaN')
#dfC['Attribute'].size
#dfC
#for i in range(0,dfC['Attribute'].size):
 #   dfC['Nvote'][i] = dfC['index'][i+1] - dfC['index'][i]

index Attribute     ID     Ig Vroot  URL  Nvote
0         294         C  10001  10001   NaN  NaN    NaN
1         298         C  10002  10002   NaN  NaN    NaN
2         301         C  10003  10003   NaN  NaN    NaN
3         305         C  10004  10004   NaN  NaN    NaN
4         307         C  10005  10005   NaN  NaN    NaN
5         309         C  10006  10006   NaN  NaN    NaN
6         312         C  10007  10007   NaN  NaN    NaN
7         314         C  10008  10008   NaN  NaN    NaN
8         316         C  10009  10009   NaN  NaN    NaN
9         319         C  10010  10010   NaN  NaN    NaN
10        326         C  10011  10011   NaN  NaN    NaN
11        332         C  10012  10012   NaN  NaN    NaN
12        335         C  10013  10013   NaN  NaN    NaN
13        337         C  10014  10014   NaN  NaN    NaN
14        339         C  10015  10015   NaN  NaN    NaN
15        341         C  10016  10016   NaN  NaN    NaN
16        344         C  10017  10017   NaN  NaN    NaN
17        349         C  10018  10018   NaN  NaN    NaN
18        351         C  10019  10019   NaN  NaN    NaN
19        363         C  10020  10020   NaN  NaN    NaN
20        368         C  10021  10021   NaN  NaN    NaN
21        386         C  10022  10022   NaN  NaN    NaN
22        390         C  10023  10023   NaN  NaN    NaN
23        392         C  10024  10024   NaN  NaN    NaN
24        394         C  10025  10025   NaN  NaN    NaN
25        396         C  10026  10026   NaN  NaN    NaN
26        398         C  10027  10027   NaN  NaN    NaN
27        402         C  10028  10028   NaN  NaN    NaN
28        404         C  10029  10029   NaN  NaN    NaN
29        406         C  10030  10030   NaN  NaN    NaN
...       ...       ...    ...    ...   ...  ...    ...
32681  131537         C  42682  42682   NaN  NaN    NaN
32682  131539         C  42683  42683   NaN  NaN    NaN
32683  131541         C  42684  42684   NaN  NaN    NaN
32684  131544         C  42685  42685   NaN  NaN    NaN
32685  131547         C  42686  42686   NaN  NaN    NaN
32686  131549         C  42687  42687   NaN  NaN    NaN
32687  131554         C  42688  42688   NaN  NaN    NaN
32688  131557         C  42689  42689   NaN  NaN    NaN
32689  131559         C  42690  42690   NaN  NaN    NaN
32690  131562         C  42691  42691   NaN  NaN    NaN
32691  131565         C  42692  42692   NaN  NaN    NaN
32692  131575         C  42693  42693   NaN  NaN    NaN
32693  131578         C  42694  42694   NaN  NaN    NaN
32694  131582         C  42695  42695   NaN  NaN    NaN
32695  131585         C  42696  42696   NaN  NaN    NaN
32696  131587         C  42697  42697   NaN  NaN    NaN
32697  131592         C  42698  42698   NaN  NaN    NaN
32698  131601         C  42699  42699   NaN  NaN    NaN
32699  131606         C  42700  42700   NaN  NaN    NaN
32700  131609         C  42701  42701   NaN  NaN    NaN
32701  131613         C  42702  42702   NaN  NaN    NaN
32702  131616         C  42703  42703   NaN  NaN    NaN
32703  131622         C  42704  42704   NaN  NaN    NaN
32704  131627         C  42705  42705   NaN  NaN    NaN
32705  131631         C  42706  42706   NaN  NaN    NaN
32706  131636         C  42707  42707   NaN  NaN    NaN
32707  131643         C  42708  42708   NaN  NaN    NaN
32708  131650         C  42709  42709   NaN  NaN    NaN
32709  131653         C  42710  42710   NaN  NaN    NaN
32710  131657         C  42711  42711   NaN  NaN    NaN

[32711 rows x 7 columns]

In [7]:
i = 0
dfC['index'][i+1] - dfC['index'][i]

4